omen2/omen.py

```diff
@@ -16,6 +26,19 @@
 T = TypeVar("T", bound=Table)
 
 
+@contextlib.contextmanager
+def _omen_tx_guard(omen: "Omen") -> Iterator[None]:
+    try:
+        with omen.db.transaction():
+            yield
+    except BaseException:
+        # Roll back in-memory values to their on-disk state
+        for tbl in omen.tables.values():
```

Comment:
in omen, transactions normally block.  ie:  other threads cannot make changes to objects that are currently being modified

with obj:
   x = 4
   y = 5

this "rolls back" on exception the way you expect

so if you're changing A, B and C, you should put A B and C in a "changeset" block

a system-wide transaction should work the same way

Reply:
if you want to do a "omen-wide-transaction", it should put each object into "with" state *as they are modified, and only if they aren't in one"

Reply:
it will also need to put the whole tables themselves into "with state", on inserts and on deletes, etc.   you can solve this either with locks or a thread-local cache of edits.   for the objects, we decided locks were better because it leads to better communication between threads (what do you do about conflicts?) , but for tables, TLS might be better?   have to think about it

Reply:
i would be ok with *ignoring* removes/adds/deletes for now, and just get updates to "auto-with" each other

Reply:
```

with o2.transaction():  
    obj1.x = 4  # (obj1 is now locked, just like a with block)
    obj1.y = 4 
    o2.tab1.add(Obj(x=3,y=4))  # (tab1 puts the new object in a thread-local cache (TLC), using __enter__)
    o2.tab1.remove(zzz)  # (tab1 marks zzz as removed in the TLC)

# tab1 executes add/removes in __exit__
# transaction calls __exit__ on all objects that it called __enter__ on